feat(orientation_filter): Added additional orientation filter options

reinzor opened the pull request:
Enables plan references with different orientations for omni-base
controllers. The following options are added:

- Backward (backward path traversal, pose points to previous point)
- Leftward (lateral path traversal in the positive y direction)
- Rightward (lateral path traversal in the negative y direction)

Also altered the orientation calculation and added additional parameter `window_size` that can be used to smoothen the orientation calculation based on the position derivative.

Commit:
feat(orientation_filter): Added additional orientation filter options

Enables plan references with different orientations for omni-base
controllers. The following options are added:

- Backward (backward path traversal, pose points to previous point)
- Leftward (lateral path traversal in the positive y direction)
- Rightward (lateral path traversal in the negative y direction)

mikeferguson commented:
@DLu -- do you have a chance to look over this? I think you're more versed in the global_planner than I am

DLu commented on global_planner/cfg/GlobalPlanner.cfg:
```diff
@@ -16,9 +16,12 @@ orientation_enum = gen.enum([
                              int_t, 3, "Forward orientation until last straightaway, then a linear blend until the goal pose"),
+    gen.const("Backward",    int_t, 4, "Orientations point to the previous point on the path"),
+    gen.const("Leftward",    int_t, 5, "Orientations point to the next point on the path (in lateral direction)"),
+    gen.const("Rightward",   int_t, 6, "Orientations point to the previous point on the path + pi / 2 (in lateral direction)"),
```
Would you mind tweaking the descriptions here such that they specify which axis is leading? i.e. Backward is the negative x axis points at the next point, and Leftward is the positive y axis (I think).

DLu commented on global_planner/src/orientation_filter.cpp:
```diff
@@ -60,6 +60,24 @@ void OrientationFilter::processPath(const geometry_msgs::PoseStamped& start,
+                pointToNext(path, i);
+                set_angle(&path[i], angles::normalize_angle(getYaw(path[i]) + M_PI_2));
+            }
+            break;
```
Nonblocking suggestion: If you wanted to get really fancy, you could have one additional `ARBITRARY_OFFSET` mode that had a dynamically reconfigurable variable that would specify what angle to offset each angle with, and then use that variable in [`pointToNext`](https://github.com/ruvu/navigation/blob/1f4cb257dc299cd51fd2f3ea866ef0e2cd382c20/global_planner/src/orientation_filter.cpp#L115). Then, when you selected LEFTWARD in the dynamic reconfigure callback, it could set the variable to -M_PI_2. 

This would also have the benefit of reducing the number of calls to quaternion conversion methods.

reinzor replied:
Thanks for the suggestion. However, I currently do not see a use-case for this.

DLu reviewed (commented):
Thanks for the PR. Let me know what you think of my suggestions, and then we can move forward.

It would be cool (though not necessary) if you had a video showing the new behaviors.

Commit:
Updated orientation filter option description

reinzor commented:
Video showing the behaviors:

https://www.youtube.com/watch?v=NQ2z90i5V2I&feature=youtu.be

Commit:
Added window size parameter to orientation filter

Previously, the orientation was calculated using the current and the
next point. However, when the path is somewhat jumpy, this results in
poor orientations. By adding this parameter and altering the orientation
calculation, the calculated orientation can be smoothened along the path
by taking into account a larger window. The orientation of index point i
will be calculated using the positions of i - window_size and i +
window_size.

reinzor commented:
Thanks, I also updated the ROS wiki of melodic

moriarty commented:
FYI: @reinzor: @rayman opened issue #760 because my PR #752 appears to have broken this 👎
